### ISSUE TITLE
Hashing functions output clarification

### ISSUE THREAD

**Issue by ben-ce** [Enzoic/enzoic-javascript-client#12]
Hi!

Could you please elaborate on the sha256crypt hashing functions output format?
The API docs don't specify the used format of these types of hashes and the ruby client seems to be doing it differently. Do you only use the default 5000 rounds of calculation with the sha256crypt hashes? Do you ever include the number of rounds in the output format?

I gathered the following:
The JavaScript client implementation for sha256crypt uses [unixcrypt](https://github.com/markusberg/unixcrypt/tree/master) and it does not cut the salt (see below for sha512crypt) and is working correctly with the different salt formats: 
- if the number of rounds is omitted the output will omit it
- if the number of rounds is specified in the salt, the output will contain it

The Ruby client implementation produces the output without the number of rounds part (`rounds=<N>$`) even if the input salt contains it. [Evidence in the testing file](https://github.com/Enzoic/enzoic-ruby-client/blob/6f19fd1906ee9495d16d82cc93092c0a378e6c86/test/hashing_test.rb#L75).
Comparing that to the [same test](https://github.com/Enzoic/enzoic-javascript-client/blob/0122c1ecd42ac87deb0bc1bf99fe333e2bc454a2/test/hashing.js#L201) with the same inputs in the JavaScript implementation, we see that they expect different outputs, the JS one with the number of rounds part and the Ruby one without the number of rounds part.

So the questions again: Do you only use the default 5000 rounds of calculation with the sha256crypt hashes? Do you ever include the number of rounds in the output format?

Another observation:
The JavaScript client implementation uses [sha512crypt-node](https://github.com/mvo5/sha512crypt-node/tree/master) for sha512crypt type hashes. The `sha512Crypt` function [here](https://github.com/Enzoic/enzoic-javascript-client/blob/0122c1ecd42ac87deb0bc1bf99fe333e2bc454a2/src/hashing.js#L260) takes the `sSalt` parameter, cuts out the magic `$6$` value at the beginning (with `sSalt.substring(3)`) of the string and passes it to the used library's exported `b64_sha512crypt` function. This will call the `b64_sha512crypt` hash function which will use the default 5000 rounds and it will produce the following format `$6$<salt>$<digest>` where `<salt>` does not contain the `rounds=<N>$` part. This is all good when using a salt value like '$6$testsalt'.

But the used library allows (as does the standard) the salt to describe the number of rounds.
When we give the enzoic sha512Crypt function a salt value where the number of rounds part is used (like '$6$rounds=4999$testsalt'), the sha512crypt-node module's function will [throw an error](https://github.com/mvo5/sha512crypt-node/blob/d487051c0ec5ef99f104da113c3f7fd529d50c75/sha512crypt.js#L140), because the enzoic function cut off the magic `$6$` before passing it (the length of the `magic_array` will be bigger than 1, but the checked item won't be equal to 6):

```
Error: Got 'rounds=5000$52450745' but only SHA512 ($6$) algorithm supported
at sha512crypt (node_modules/sha512crypt-node/sha512crypt.js:133:19)
at Object.sha512Crypt (src/hashing.js:261:16)
at Context.<anonymous> (test/hashing.js:210:28)
at process.processImmediate (node:internal/timers:478:21)
```

Wouldn't it be better if it did not cut the salt at the magic id value and let the underlying library do the parsing of the salt?

**Comment by mbparker** [comment]
Thank you for your report, @ben-ce 

We'll review all this and get back to you on your questions.

**Comment by mwilson71** [comment]
hi @ben-ce,

Thanks for reaching out and reviewing our code.

In terms of what output is expected, it's really dictated by the contents of breaches we pick up that contain these types of hashes.  In this case, the breaches we've picked up to date have all been using the default number of hashing rounds and did not include the rounds specifier, so this probably wasn't tested adequately in all the different SDKs.  It appears typical behavior across just about every lib we've tested other than the one we use for the Ruby SDK is to include the rounds specifier in the output hash string if a rounds spec was included in the input, regardless of whether it was the default of 5000.  We've updated the behavior of the Ruby SDK to match this.  

It also appeared unnecessary to use the separate library we were using in the JavaScript SDK for SHA512 Crypt.  Instead we dropped that and switched it to use the same unixcrypt lib we already used for SHA256 Crypt.  This fixed the behavior you pointed out with not supporting the rounds parameter.  

We also noticed the password type for SHA256 Crypt was erroneous and did not match our infrastructure (it should have been type 43, not 41) and support for HMAC SHA1 (which is what 41 should have been) was absent.  These issues have both been corrected in our latest SDK releases.

Thanks for the heads up!
